File: ios-native/ServiceCrewTech/Models/Job.swift
import Foundation

// MARK: - Job Model

struct Job: Identifiable, Codable, Hashable {
    let id: String
    let jobNumber: String
    let title: String?
    let description: String?
    let status: JobStatus
    let priority: JobPriority
    let tradeType: TradeType?

    var displayTitle: String {
        title ?? "Untitled Job"
    }
    let scheduledStart: Date?
    let scheduledEnd: Date?
    let timeWindowStart: String?
    let timeWindowEnd: String?
    let estimatedDuration: Int?
    let actualDuration: Int?
    let customer: Customer?
    let property: Property?
    let serviceType: ServiceType?
    var notes: [JobNote]?
    let lineItems: [LineItem]?
    var photos: [JobPhoto]?
    let createdAt: Date?
    let updatedAt: Date?
    let assignments: [JobAssignment]?

    // Helper computed properties for non-optional access
    var safeNotes: [JobNote] { notes ?? [] }
    var safeLineItems: [LineItem] { lineItems ?? [] }
    var safePhotos: [JobPhoto] { photos ?? [] }

    // Get assigned technician name
    var assignedTechnicianName: String? {
        guard let assignments = assignments, let firstAssignment = assignments.first,
              let technician = firstAssignment.technician else { return nil }
        return technician.displayName
    }
}

// MARK: - Job Assignment

struct JobAssignment: Codable, Hashable {
    let id: String
    let status: String?
    let technician: AssignedTechnician?
}

struct AssignedTechnician: Codable, Hashable {
    let id: String
    let user: TechnicianUserInfo?

    var displayName: String {
        guard let user = user else { return "Unknown" }
        let firstName = user.firstName ?? ""
        let lastName = user.lastName ?? ""
        let name = "\(firstName) \(lastName)".trimmingCharacters(in: .whitespaces)
        return name.isEmpty ? "Unknown" : name
    }
}

struct TechnicianUserInfo: Codable, Hashable {
    let firstName: String?
    let lastName: String?
}

// MARK: - Job Status

enum JobStatus: String, Codable, CaseIterable {
    case pending = "PENDING"
    case scheduled = "SCHEDULED"
    case dispatched = "DISPATCHED"
    case enRoute = "EN_ROUTE"
    case inProgress = "IN_PROGRESS"
    case onHold = "ON_HOLD"
    case completed = "COMPLETED"
    case cancelled = "CANCELLED"

    var displayName: String {
        switch self {
        case .pending: return "Pending"
        case .scheduled: return "Scheduled"
        case .dispatched: return "Dispatched"
        case .enRoute: return "En Route"
        case .inProgress: return "In Progress"
        case .onHold: return "On Hold"
        case .completed: return "Completed"
        case .cancelled: return "Cancelled"
        }
    }

    var color: String {
        switch self {
        case .pending: return "gray"
        case .scheduled: return "gray"
        case .dispatched: return "blue"
        case .enRoute: return "purple"
        case .inProgress: return "orange"
        case .onHold: return "yellow"
        case .completed: return "green"
        case .cancelled: return "red"
        }
    }

    var isActive: Bool {
        switch self {
        case .inProgress, .enRoute: return true
        default: return false
        }
    }
}

// MARK: - Job Priority

enum JobPriority: String, Codable, CaseIterable {
    case low = "LOW"
    case normal = "NORMAL"
    case high = "HIGH"
    case urgent = "URGENT"
    case emergency = "EMERGENCY"

    var displayName: String {
        rawValue.capitalized
    }

    var color: String {
        switch self {
        case .low: return "gray"
        case .normal: return "blue"
        case .high: return "orange"
        case .urgent: return "red"
        case .emergency: return "purple"
        }
    }
}

// MARK: - Trade Type

enum TradeType: String, Codable, CaseIterable {
    case hvac = "HVAC"
    case plumbing = "PLUMBING"
    case electrical = "ELECTRICAL"
    case appliance = "APPLIANCE"
    case generalContractor = "GENERAL_CONTRACTOR"
    case other = "OTHER"

    var displayName: String {
        switch self {
        case .hvac: return "HVAC"
        case .plumbing: return "Plumbing"
        case .electrical: return "Electrical"
        case .appliance: return "Appliance"
        case .generalContractor: return "General Contractor"
        case .other: return "Other"
        }
    }

    var iconName: String {
        switch self {
        case .hvac: return "fan.fill"
        case .plumbing: return "drop.fill"
        case .electrical: return "bolt.fill"
        case .appliance: return "washer.fill"
        case .generalContractor: return "hammer.fill"
        case .other: return "wrench.fill"
        }
    }
}

// MARK: - Job Note

struct JobNote: Identifiable, Codable, Hashable {
    let id: String
    let content: String
    let noteType: NoteType
    let createdAt: Date
    let createdBy: String?

    enum NoteType: String, Codable {
        case technician = "TECHNICIAN"
        case dispatcher = "DISPATCHER"
        case customer = "CUSTOMER"
        case system = "SYSTEM"
    }
}

// MARK: - Line Item

struct LineItem: Identifiable, Codable, Hashable {
    let id: String
    let description: String
    let itemType: ItemType?

    // Prisma Decimal fields
    private let quantityRaw: AnyCodableNumber?
    private let unitPriceRaw: AnyCodableNumber?
    private let totalPriceRaw: AnyCodableNumber?

    var quantity: Double { quantityRaw?.doubleValue ?? 0 }
    var unitPrice: Double { unitPriceRaw?.doubleValue ?? 0 }
    var totalPrice: Double { totalPriceRaw?.doubleValue ?? 0 }

    private enum CodingKeys: String, CodingKey {
        case id, description, itemType
        case quantityRaw = "quantity"
        case unitPriceRaw = "unitPrice"
        case totalPriceRaw = "totalPrice"
    }

    enum ItemType: String, Codable {
        case labor = "LABOR"
        case part = "PART"
        case material = "MATERIAL"
        case service = "SERVICE"
        case other = "OTHER"
    }
}

// MARK: - Job Photo

struct JobPhoto: Identifiable, Codable, Hashable {
    let id: String
    let url: String
    let caption: String?
    let photoType: PhotoType
    let latitude: Double?
    let longitude: Double?
    let takenAt: Date

    enum PhotoType: String, Codable {
        case before = "BEFORE"
        case during = "DURING"
        case after = "AFTER"
        case equipment = "EQUIPMENT"
        case damage = "DAMAGE"
        case other = "OTHER"
    }
}

// MARK: - Service Type

struct ServiceType: Codable, Hashable {
    let id: String
    let name: String
    let description: String?
}

// MARK: - Job Extensions

extension Job {
    var customerName: String {
        guard let customer = customer else { return "Unknown" }
        let firstName = customer.firstName ?? ""
        let lastName = customer.lastName ?? ""
        let name = "\(firstName) \(lastName)".trimmingCharacters(in: .whitespaces)
        return name.isEmpty ? "Unknown" : name
    }

    var fullAddress: String {
        guard let property = property else { return "No address" }
        return "\(property.address), \(property.city), \(property.state) \(property.zip)"
    }

    var shortAddress: String {
        guard let property = property else { return "No address" }
        return "\(property.address), \(property.city)"
    }

    var canStart: Bool {
        [.scheduled, .dispatched].contains(status)
    }

    var canPause: Bool {
        status == .inProgress
    }

    var canComplete: Bool {
        status == .inProgress
    }

    var partsUsed: [LineItem] {
        safeLineItems.filter { $0.itemType == .part }
    }

    var laborItems: [LineItem] {
        safeLineItems.filter { $0.itemType == .labor }
    }
}

File: flutter_app/lib/models/job.dart
import 'package:json_annotation/json_annotation.dart';
import 'customer.dart';

part 'job.g.dart';

@JsonSerializable()
class Job {
  final String id;
  final String jobNumber;
  final String title;
  final String? description;
  final JobStatus status;
  final JobPriority priority;
  final TradeType tradeType;
  final DateTime? scheduledStart;
  final DateTime? scheduledEnd;
  final String? timeWindowStart;
  final String? timeWindowEnd;
  final int? estimatedDuration;
  final int? actualDuration;
  final Customer? customer;
  final Property? property;
  final ServiceType? serviceType;
  final List<JobNote> notes;
  final List<LineItem> lineItems;
  final List<JobPhoto> photos;
  final DateTime createdAt;
  final DateTime updatedAt;

  Job({
    required this.id,
    required this.jobNumber,
    required this.title,
    this.description,
    required this.status,
    required this.priority,
    required this.tradeType,
    this.scheduledStart,
    this.scheduledEnd,
    this.timeWindowStart,
    this.timeWindowEnd,
    this.estimatedDuration,
    this.actualDuration,
    this.customer,
    this.property,
    this.serviceType,
    this.notes = const [],
    this.lineItems = const [],
    this.photos = const [],
    required this.createdAt,
    required this.updatedAt,
  });

  factory Job.fromJson(Map<String, dynamic> json) => _$JobFromJson(json);
  Map<String, dynamic> toJson() => _$JobToJson(this);

  String get customerName {
    if (customer == null) return 'Unknown';
    final firstName = customer!.firstName ?? '';
    final lastName = customer!.lastName ?? '';
    final name = '$firstName $lastName'.trim();
    return name.isEmpty ? 'Unknown' : name;
  }

  String get fullAddress {
    if (property == null) return 'No address';
    return '${property!.address}, ${property!.city}, ${property!.state} ${property!.zip}';
  }

  String get shortAddress {
    if (property == null) return 'No address';
    return '${property!.address}, ${property!.city}';
  }

  bool get canStart => [JobStatus.scheduled, JobStatus.dispatched].contains(status);
  bool get canPause => status == JobStatus.inProgress;
  bool get canComplete => status == JobStatus.inProgress;

  List<LineItem> get partsUsed => lineItems.where((i) => i.itemType == ItemType.part).toList();
}

enum JobStatus {
  @JsonValue('SCHEDULED')
  scheduled,
  @JsonValue('DISPATCHED')
  dispatched,
  @JsonValue('EN_ROUTE')
  enRoute,
  @JsonValue('IN_PROGRESS')
  inProgress,
  @JsonValue('ON_HOLD')
  onHold,
  @JsonValue('COMPLETED')
  completed,
  @JsonValue('CANCELLED')
  cancelled;

  String get displayName {
    switch (this) {
      case JobStatus.scheduled:
        return 'Scheduled';
      case JobStatus.dispatched:
        return 'Dispatched';
      case JobStatus.enRoute:
        return 'En Route';
      case JobStatus.inProgress:
        return 'In Progress';
      case JobStatus.onHold:
        return 'On Hold';
      case JobStatus.completed:
        return 'Completed';
      case JobStatus.cancelled:
        return 'Cancelled';
    }
  }

  String get value {
    switch (this) {
      case JobStatus.scheduled:
        return 'SCHEDULED';
      case JobStatus.dispatched:
        return 'DISPATCHED';
      case JobStatus.enRoute:
        return 'EN_ROUTE';
      case JobStatus.inProgress:
        return 'IN_PROGRESS';
      case JobStatus.onHold:
        return 'ON_HOLD';
      case JobStatus.completed:
        return 'COMPLETED';
      case JobStatus.cancelled:
        return 'CANCELLED';
    }
  }
}

enum JobPriority {
  @JsonValue('LOW')
  low,
  @JsonValue('NORMAL')
  normal,
  @JsonValue('HIGH')
  high,
  @JsonValue('URGENT')
  urgent,
  @JsonValue('EMERGENCY')
  emergency;

  String get displayName => name.toUpperCase();
}

enum TradeType {
  @JsonValue('HVAC')
  hvac,
  @JsonValue('PLUMBING')
  plumbing,
  @JsonValue('ELECTRICAL')
  electrical,
  @JsonValue('APPLIANCE')
  appliance,
  @JsonValue('GENERAL_CONTRACTOR')
  generalContractor,
  @JsonValue('OTHER')
  other;

  String get displayName {
    switch (this) {
      case TradeType.hvac:
        return 'HVAC';
      case TradeType.plumbing:
        return 'Plumbing';
      case TradeType.electrical:
        return 'Electrical';
      case TradeType.appliance:
        return 'Appliance';
      case TradeType.generalContractor:
        return 'General Contractor';
      case TradeType.other:
        return 'Other';
    }
  }
}

@JsonSerializable()
class JobNote {
  final String id;
  final String content;
  final NoteType noteType;
  final DateTime createdAt;
  final String? createdBy;

  JobNote({
    required this.id,
    required this.content,
    required this.noteType,
    required this.createdAt,
    this.createdBy,
  });

  factory JobNote.fromJson(Map<String, dynamic> json) => _$JobNoteFromJson(json);
  Map<String, dynamic> toJson() => _$JobNoteToJson(this);
}

enum NoteType {
  @JsonValue('TECHNICIAN')
  technician,
  @JsonValue('DISPATCHER')
  dispatcher,
  @JsonValue('CUSTOMER')
  customer,
  @JsonValue('SYSTEM')
  system;
}

@JsonSerializable()
class LineItem {
  final String id;
  final String description;
  final double quantity;
  final double unitPrice;
  final double totalPrice;
  final ItemType itemType;

  LineItem({
    required this.id,
    required this.description,
    required this.quantity,
    required this.unitPrice,
    required this.totalPrice,
    required this.itemType,
  });

  factory LineItem.fromJson(Map<String, dynamic> json) => _$LineItemFromJson(json);
  Map<String, dynamic> toJson() => _$LineItemToJson(this);
}

enum ItemType {
  @JsonValue('LABOR')
  labor,
  @JsonValue('PART')
  part,
  @JsonValue('MATERIAL')
  material,
  @JsonValue('SERVICE')
  service,
  @JsonValue('OTHER')
  other;
}

@JsonSerializable()
class JobPhoto {
  final String id;
  final String url;
  final String? caption;
  final PhotoType photoType;
  final double? latitude;
  final double? longitude;
  final DateTime takenAt;

  JobPhoto({
    required this.id,
    required this.url,
    this.caption,
    required this.photoType,
    this.latitude,
    this.longitude,
    required this.takenAt,
  });

  factory JobPhoto.fromJson(Map<String, dynamic> json) => _$JobPhotoFromJson(json);
  Map<String, dynamic> toJson() => _$JobPhotoToJson(this);
}

enum PhotoType {
  @JsonValue('BEFORE')
  before,
  @JsonValue('DURING')
  during,
  @JsonValue('AFTER')
  after,
  @JsonValue('EQUIPMENT')
  equipment,
  @JsonValue('DAMAGE')
  damage,
  @JsonValue('OTHER')
  other;
}

@JsonSerializable()
class ServiceType {
  final String id;
  final String name;
  final String? description;

  ServiceType({
    required this.id,
    required this.name,
    this.description,
  });

  factory ServiceType.fromJson(Map<String, dynamic> json) => _$ServiceTypeFromJson(json);
  Map<String, dynamic> toJson() => _$ServiceTypeToJson(this);
}
